### PR TITLE
Add utility to retrieve user's home directory

### LINKED ISSUE #60: Implement `homedir`
See [utils-homedir](https://github.com/kgryte/utils-homedir). `@stdlib/utils/homedir`.

### PR DESCRIPTION
**Pull request by Planeshifter**
Resolves #60 .

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

* [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
* [x] Read and understand the [Code of Conduct][code-of-conduct].
* [x] Read and understood the [licensing terms][license].
* [x] Searched for existing issues and pull requests __before__ submitting this pull request.
* [x] Filed an issue (or an issue already existed) __prior to__ submitting this pull request.
* [x] Rebased onto latest `develop`.
* [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

* contains a straight-forward port of [utils-homedir](https://github.com/kgryte/utils-homedir) to `stdlib`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

* resolves #
* fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

---

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/master/CONTRIBUTING.md
[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/master/CODE_OF_CONDUCT.md
[license]: https://github.com/stdlib-js/stdlib/blob/master/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+<!-- /.intro -->
+
+<!-- Package usage documentation. -->
+
+<section class="usage">
+
+## Usage
+
+``` javascript
+var homedir = require( '@stdlib/lib/node_modules/@stdlib/utils/homedir' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md]: The require statement is incorrect.

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+Returns the current user's `home` directory.
+
+``` javascript
+var home = homedir();
+// e.g., returns '/Users/<username>'
+```
+
+If unable to locate a `home` directory, the module returns `null`.
+
+``` javascript
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md]: I am not sure we should include this example. Could cause issues if we ever perform automated testing of doc examples.

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+<!-- /.usage -->
+
+<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="notes">
+
+## Notes
+
+* This module primarily checks various [environment variables][environment-variables] to locate a `home` directory. Note that this approach has __security vulnerabilities__, as attackers can tamper with [environment variables][environment-variables].
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md]: > The implementation primarily checks...

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+<!-- /.notes -->
+
+<!-- Package usage examples. -->
+
+<section class="examples">
+
+## Examples
+
+``` javascript
+var homedir = require( '@stdlib/lib/node_modules/@stdlib/utils/homedir' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md]: Require statement is incorrect.

File: lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+	env = process.env;
+	process.env = {};
+
+	t.strictEqual( homedir(), null, 'returns null' );
+	process.env = env;
+	t.end();
+});
+
+tape( 'the function returns the `/root` directory if run as `root` in a linux environment', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js]: All instances of `linux` => `Linux` in string descriptions.

File: lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+	process.env = {
+		'LNAME': 'beep'
+	};
+
+	t.strictEqual( homedir(), '/home/beep', 'returns home directory' );
+	process.env = env;
+	t.end();
+});
+
+tape( 'the function returns a home directory in a Mac OSX environment (USERNAME)', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js]: All instances of `OSX` => `OS X` in string descriptions.

File: lib/node_modules/@stdlib/utils/homedir/lib/homedir.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	// https://github.com/libuv/libuv/blob/9fbcca048181b927cfcdb5c6c49e5bdff173aad5/src/unix/core.c#L1030
+	home = env[ 'HOME' ];
+	if ( home ) {
+		return home;
+	}
+	// Get the current user account (https://docs.python.org/2/library/getpass.html):
+	user = env[ 'LOGNAME' ] || env[ 'USER' ] || env[ 'LNAME' ] || env[ 'USERNAME' ];
+
+	// If on Mac OSX, use the Mac path convention (http://apple.stackexchange.com/questions/119230/what-is-standard-for-os-x-filesystem-e-g-opt-vs-usr)...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/lib/homedir.js]: OS X

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+console.log( homedir() );
+```
+
+</section>
+
+<!-- /.examples -->
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md:66]: Missing CLI docs.

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+#### homedir()
+
+Returns the current user's `home` directory.
+
+``` javascript
+var home = homedir();
+// e.g., returns '/Users/<username>'
+```
+
+If unable to locate a `home` directory, the module returns `null`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md]: > ...the function returns `null`.

**Review by kgryte — CHANGES_REQUESTED** [review]

**Comment by Planeshifter** [comment]
Could you take another look?

File: lib/node_modules/@stdlib/utils/homedir/README.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="references">
+
+</section>
+
+<!-- /.references -->
+
+
+---
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/README.md:68]: CLI docs go above the "references" section.

File: lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+tape( 'main export is a function', function test( t ) {
+	var homedir;
+	t.ok( true, __filename );
+	homedir = require( mpath );
+	t.strictEqual( typeof homedir, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function supports older Node versions', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/test/test.homedir.js]: How does this test whether the implementation supports older Node versions? Considering how this package works, this should be in the other test file.

**Review by kgryte — CHANGES_REQUESTED** [review]

File: lib/node_modules/@stdlib/utils/homedir/test/test.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+tape( 'the function supports older Node versions', function test( t ) {
+	var homedir;
+	var opts;
+	var env;
+
+	opts = {
+		'os': {
+			'homedir': void 0
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/homedir/test/test.js:48]: Can you confirm that this works as intended? Possible that `proxyquire` ignores `homedir` when set to `undefined`, falling back to the original implementation. If so, can set to `false`.

This should be apparent when generating a coverage report and whether all code branches are covered (notably within the main `index.js` file).

**Review by kgryte — COMMENTED** [review]
Need to confirm test works as intended.

File: lib/node_modules/@stdlib/utils/homedir/test/test.js (hunk 12 above)
Comment (Planeshifter) [reply lib/node_modules/@stdlib/utils/homedir/test/test.js:48]: Yes, can confirm that we reach 100% code coverage for `homedir`.

Comment (kgryte) [reply lib/node_modules/@stdlib/utils/homedir/test/test.js:48]: Including branch coverage?

Comment (Planeshifter) [reply lib/node_modules/@stdlib/utils/homedir/test/test.js:48]: Yes.

Comment (kgryte) [reply lib/node_modules/@stdlib/utils/homedir/test/test.js:48]: Must be because `os` does get mocked with an empty `object`. Makes sense.

**Pull request merged**